tests/x11/thunderbird/thunderbird_common.pm

```diff
@@ -0,0 +1,171 @@
+    my $mail_user       = $config->{$account}->{user};
+    my $mail_passwd     = $config->{$account}->{passwd};
+    my $mail_sendport   = $config->{$account}->{sendport};
+    my $port_key        = $proto eq 'pop' ? 'recvport' : 'imapport';
+    my $mail_recvport   = $config->{$account}->{$port_key};
+
+    $self->tb_start($mail_box, $mail_passwd);
+
+    send_key "alt-n";
+    type_string "SUSE Test";
+    send_key "alt-e";
+    wait_screen_change { type_string "$mail_box" };
+    send_key "alt-p";
+    wait_screen_change { type_string "$mail_passwd" };
+
+    send_key "alt-c";
+    assert_screen "thunderbird_wizard-discovery-ok";
+
+    assert_and_click "thunderbird_wizard-set-pop" if ($proto eq 'pop');
+    assert_screen "thunderbird_wizard-$proto-selected";
+    # done
+    send_key "alt-d";
+    # workaround: for some reasons, when using self signed cert thunderbird doesn't accept the password during setup
+    # but does accept it later on. So, we have to click 'done', delete the password, click 'done' again and we are good to go.
+    # Of course, we will have to input the password later on (eg when clicking on 'Inbox')
+    send_key "alt-p";
+    send_key "delete";
+    send_key "alt-d";
+
+    # skip additional integrations
+    assert_and_click "thunderbird_skip-system-integration";
+
+    # confirm that we accept the self signed cert
+    assert_screen "thunderbird_incoming-self-signed-cert";
+    send_key "alt-c";
+
+    assert_screen "thunderbird_account-configured-mainwin";
+    assert_and_click "thunderbird_account-inbox";
+
+    if ($proto eq 'pop') {
+	    # when using POP simply clicking on inbox doesn't show the dialog password, so we have to click on 'Get Messages'
+	    assert_and_click "thunderbird_get-messages";
+    }
+
+    # we now have to input and store the imap password
+    assert_and_click "thunderbird_passwd-diag-use-password-manager";
+    assert_and_click "thunderbird_passwd-diag-entry";
+    wait_screen_change { type_string "$mail_passwd" };
+    send_key "ret";
+
+    assert_screen "thunder_account-inbox-overview";
+}
+
+sub tb_send_message {
+    my ($self, $account) = @_;
+
+    my $config       = $self->getconfig_emailaccount;
+    my $mailbox      = $config->{$account}->{mailbox};
+    my $mail_passwd  = $config->{$account}->{passwd};
+    my $mail_subject = $self->get_dated_random_string(4);
+
+    send_key "ctrl-m";
+    assert_screen "thunderbird_mail-compose-message";
+    wait_screen_change { type_string "$mailbox" };
+
+    send_key "alt-s"; 
+    wait_screen_change { type_string "$mail_subject this is a test mail" };
+
+    send_key "tab";
+    type_string "Test email send and receive.";
+    send_key "ctrl-ret";
+    # we can't use "ret" because it doesn't always work
+    assert_and_click "thunderbird_really-send-message";
+
+    # it may happen that the 'send error' dialog is shown before the 'confirm certificate' dialog
+    if (check_screen "thunderbird_send-message-error", 3) {
```

Comment:
Please use a multi-tag `assert_screen` with `match_has_tag` instead of `check_screen` with non-zero timeout to prevent introducing any timing dependant behaviour, to save test execution time as well as state more explicitly from the testers point of view what are the expected alternatives. For example:

```perl
assert_screen([qw(yast2_console-finished yast2_missing_package)]);
if (match_has_tag('yast2_missing_package')) {
    send_key 'alt-o';  # confirm package installation
    assert_screen 'yast2_console-finished';
}
```

Reply:
thanks for tips, will update the code :) :+1: